File: leetcode/max-difference-you-can-get-from-changing-an-integer/dart/solution1/solution.dart
class Solution {
  int maxDiff(int n) {
    final digits = n.toDigits().toList();

    int maxFrom = digits.first;
    const maxTo = 9;

    for (final d in digits) {
      if (d == 9) continue;

      maxFrom = d;
      break;
    }

    int minFrom = digits.first;
    int minTo = 1;

    for (final (i, d) in digits.indexed) {
      if (d == 0) continue;
      if (d == 1) continue;

      minFrom = d;

      if (i != 0) {
        minTo = 0;
      }

      break;
    }

    return digits.map((e) => e == maxFrom ? maxTo : e).mountToInt() -
        digits.map((e) => e == minFrom ? minTo : e).mountToInt();
  }
}

int? largestPowerOfOrNull(int x, {int base = 10}) {
  try {
    return largestPowerOf(x, base: base);
  } on Exception {
    return null;
  }
}

int largestPowerOf(int x, {int base = 10}) {
  final (_, power) = largestExponentAndPowerOf(x, base: base);
  return power;
}

(int exponent, int power) largestExponentAndPowerOf(int x, {int base = 10}) {
  int exponent = 0;
  int power = 1;

  while (power * base <= x) {
    power = power * base;
    exponent = exponent + 1;
  }

  return (exponent, power);
}

extension DigitOperations on int {
  Iterable<int> toDigits({int radix = 10}) sync* {
    int self = abs();

    int power = largestPowerOfOrNull(self, base: radix) ?? 1;

    do {
      yield self ~/ power;
      self = self % power;
      power = power ~/ radix;
    } while (power > 0);
  }
}

extension DigitOperationsIterable<T extends Iterable<int>> on T {
  int mountToInt({int radix = 10}) {
    int mounted = 0;

    for (final digit in this) {
      mounted = mounted * radix;
      mounted = mounted + digit.abs();
    }

    return mounted;
  }
}

File: leetcode/max-difference-you-can-get-from-changing-an-integer/python3/solution1/solution.py
from typing import Iterator, Iterable, Optional, Tuple


class Solution:
    def maxDiff(self, n: int) -> int:
        digits = list(to_digits(n))
        
        max_from = digits[0]
        max_to = 9
        for d in digits:
            if d == 9:
                continue
            max_from = d
            break
        
        min_from = digits[0]
        min_to = 1
        for i, d in enumerate(digits):
            if d == 0:
                continue
            if d == 1:
                continue
            min_from = d
            if i != 0:
                min_to = 0
            break
        
        max_num = mount_to_int([max_to if e == max_from else e for e in digits])
        min_num = mount_to_int([min_to if e == min_from else e for e in digits])
        
        return max_num - min_num


def largest_power_of_or_none(x: int, base: int = 10) -> Optional[int]:
    try:
        return largest_power_of(x, base)
    except Exception:
        return None


def largest_power_of(x: int, base: int = 10) -> int:
    _, power = largest_exponent_and_power_of(x, base)
    return power


def largest_exponent_and_power_of(x: int, base: int = 10) -> Tuple[int, int]:
    exponent = 0
    power = 1
    while power * base <= x:
        power *= base
        exponent += 1
    return exponent, power


def to_digits(n: int, radix: int = 10) -> Iterator[int]:
    self = abs(n)
    power = largest_power_of_or_none(self, radix)
    if power is None:
        power = 1
    
    while power > 0:
        yield self // power
        self %= power
        power //= radix


def mount_to_int(digits: Iterable[int], radix: int = 10) -> int:
    mounted = 0
    for digit in digits:
        mounted *= radix
        mounted += abs(digit)
    return mounted
